File: fastapi_apps/main_entry/app/routes/provider_connection.py
"""Provider Connection Health Check Routes.

Exposes provider connection health check endpoints.
Uses provider_api_getters for API token resolution and health checking.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from provider_api_getters import (
    ProviderHealthChecker,
    check_provider_connection,
    PROVIDER_REGISTRY,
)


class ProviderConnectionResponse(BaseModel):
    """Provider connection health check response."""

    provider: str
    status: str
    latency_ms: Optional[float] = None
    message: Optional[str] = None
    error: Optional[str] = None
    timestamp: str
    config_used: Optional[Dict[str, Any]] = None


class ProvidersListResponse(BaseModel):
    """List of available providers."""

    providers: List[str]
    count: int
    timestamp: str


class RuntimeOverrideRequest(BaseModel):
    """Request body for runtime proxy/client override testing."""

    proxy: Optional[Dict[str, Any]] = None
    client: Optional[Dict[str, Any]] = None
    headers: Optional[Dict[str, str]] = None


router = APIRouter()


def get_static_config():
    """Get static config from main module."""
    from app.main import static_config_env
    return static_config_env


@router.get("", response_model=ProvidersListResponse)
async def list_providers() -> ProvidersListResponse:
    """
    List all available providers that can be health-checked.

    Returns a list of provider names that can be passed to the
    /{provider_name} endpoint.
    """
    providers = list(PROVIDER_REGISTRY.keys())
    unique_providers = sorted(set(providers))

    return ProvidersListResponse(
        providers=unique_providers,
        count=len(unique_providers),
        timestamp=datetime.now().isoformat(),
    )


@router.get("/{provider_name}", response_model=ProviderConnectionResponse)
async def check_provider(
    provider_name: str,
    static_config=Depends(get_static_config),
) -> ProviderConnectionResponse:
    """
    Check connection to a specific provider.

    Args:
        provider_name: The name of the provider to check
            (e.g., 'figma', 'github', 'jira', 'postgres', 'redis')

    Returns:
        Connection status including latency, success message or error details,
        and the effective configuration used for the connection test.

    Status values:
        - connected: Successfully connected to the provider
        - error: Failed to connect (check error field for details)
        - not_implemented: Provider is a placeholder (not yet implemented)
    """
    checker = ProviderHealthChecker(static_config)
    result = await checker.check(provider_name)

    return ProviderConnectionResponse(
        provider=result.provider,
        status=result.status,
        latency_ms=result.latency_ms,
        message=result.message,
        error=result.error,
        timestamp=result.timestamp,
        config_used=result.config_used,
    )


@router.post("/{provider_name}", response_model=ProviderConnectionResponse)
async def check_provider_with_override(
    provider_name: str,
    override: RuntimeOverrideRequest,
    static_config=Depends(get_static_config),
) -> ProviderConnectionResponse:
    """
    Check connection to a provider with runtime proxy/client override.

    Useful for testing VPN/proxy configurations without modifying YAML.
    The override is deep-merged with the static config (global + overwrite_root_config).

    Args:
        provider_name: The name of the provider to check
        override: Runtime override for proxy, client, and headers settings

    Example request body:
    ```json
    {
        "proxy": {
            "default_environment": "prod",
            "proxy_urls": {"prod": "http://proxy.internal:8080"},
            "cert_verify": false
        },
        "client": {
            "timeout_seconds": 120.0
        }
    }
    ```

    Returns:
        Connection status with the effective configuration used.
    """
    runtime_override = override.model_dump(exclude_none=True)
    checker = ProviderHealthChecker(static_config, runtime_override=runtime_override)
    result = await checker.check(provider_name)

    return ProviderConnectionResponse(
        provider=result.provider,
        status=result.status,
        latency_ms=result.latency_ms,
        message=result.message,
        error=result.error,
        timestamp=result.timestamp,
        config_used=result.config_used,
    )

File: fastify_apps/main_entry/src/routes/provider_connection.mjs
/**
 * Provider Connection Health Check Routes Plugin
 *
 * Exposes provider connection health check endpoints.
 * Uses provider_api_getters for API token resolution and health checking.
 */

import fastifyPlugin from 'fastify-plugin';
import {
  ProviderHealthChecker,
  PROVIDER_REGISTRY,
} from '@internal/provider-api-getters';

/**
 * Provider Connection Routes Plugin
 */
async function providerConnectionRoutesPlugin(fastify, options) {
  const prefix = options.prefix || '/healthz/providers/connection';

  // Get static config from decorator (set in main.mjs)
  const getStaticConfig = () => fastify.staticConfig;

  /**
   * GET /healthz/providers/connection
   * Returns list of available providers
   */
  fastify.get(prefix, async (request, reply) => {
    const providers = Object.keys(PROVIDER_REGISTRY);
    const uniqueProviders = [...new Set(providers)].sort();

    return {
      providers: uniqueProviders,
      count: uniqueProviders.length,
      timestamp: new Date().toISOString(),
    };
  });

  /**
   * GET /healthz/providers/connection/:providerName
   * Check connection to a specific provider
   * Returns config_used with the effective configuration used for the test
   *
   * Query Parameters (proxy overrides):
   *   proxy_env: Override default_environment (e.g., 'prod', 'dev')
   *   proxy_url: Override proxy URL for the current environment
   *   cert_verify: Override SSL certificate verification (true/false)
   *   ca_bundle: Override CA bundle path
   *   http_proxy: Override HTTP agent proxy
   *   https_proxy: Override HTTPS agent proxy
   *
   * Query Parameters (client overrides):
   *   timeout_seconds: Override request timeout in seconds
   *   timeout_ms: Override request timeout in milliseconds
   *
   * Examples:
   *   /healthz/providers/connection/gemini?proxy_env=prod&proxy_url=http://proxy:8080
   *   /healthz/providers/connection/jira?cert_verify=false&timeout_seconds=120
   *   /healthz/providers/connection/github?http_proxy=http://squid:3128
   */
  fastify.get(`${prefix}/:providerName`, async (request, reply) => {
    const { providerName } = request.params;
    const staticConfig = getStaticConfig();

    // Extract query params for proxy/client overrides
    const {
      proxy_env,
      proxy_url,
      cert_verify,
      ca_bundle,
      http_proxy,
      https_proxy,
      timeout_seconds,
      timeout_ms,
    } = request.query;

    // Build runtime override from query params
    const runtimeOverride = {};

    // Proxy overrides
    const proxyOverride = {};
    if (proxy_env !== undefined) {
      proxyOverride.default_environment = proxy_env;
    }
    if (proxy_url !== undefined) {
      const envKey = proxy_env || 'default';
      proxyOverride.proxy_urls = { [envKey]: proxy_url };
    }
    if (cert_verify !== undefined) {
      proxyOverride.cert_verify = cert_verify === 'true' || cert_verify === true;
    }
    if (ca_bundle !== undefined) {
      proxyOverride.ca_bundle = ca_bundle;
    }
    if (http_proxy !== undefined || https_proxy !== undefined) {
      proxyOverride.agent_proxy = {};
      if (http_proxy !== undefined) {
        proxyOverride.agent_proxy.http_proxy = http_proxy || null;
      }
      if (https_proxy !== undefined) {
        proxyOverride.agent_proxy.https_proxy = https_proxy || null;
      }
    }

    if (Object.keys(proxyOverride).length > 0) {
      runtimeOverride.proxy = proxyOverride;
    }

    // Client overrides
    const clientOverride = {};
    if (timeout_seconds !== undefined) {
      clientOverride.timeout_seconds = parseFloat(timeout_seconds);
    }
    if (timeout_ms !== undefined) {
      clientOverride.timeout_ms = parseInt(timeout_ms, 10);
    }

    if (Object.keys(clientOverride).length > 0) {
      runtimeOverride.client = clientOverride;
    }

    // Create checker with optional runtime override
    const hasOverride = Object.keys(runtimeOverride).length > 0;
    const checker = new ProviderHealthChecker(
      staticConfig,
      hasOverride ? runtimeOverride : null
    );
    const result = await checker.check(providerName);

    return {
      provider: result.provider,
      status: result.status,
      latency_ms: result.latency_ms,
      message: result.message,
      error: result.error,
      timestamp: result.timestamp,
      config_used: result.config_used,
    };
  });

  /**
   * POST /healthz/providers/connection/:providerName
   * Check connection with runtime proxy/client override
   *
   * Useful for testing VPN/proxy configurations without modifying YAML.
   * The override is deep-merged with the static config (global + overwrite_root_config).
   *
   * Request body:
   * {
   *   "proxy": {
   *     "default_environment": "prod",
   *     "proxy_urls": {"prod": "http://proxy.internal:8080"},
   *     "cert_verify": false
   *   },
   *   "client": {
   *     "timeout_seconds": 120.0
   *   },
   *   "headers": {
   *     "X-Custom-Header": "value"
   *   }
   * }
   */
  fastify.post(`${prefix}/:providerName`, async (request, reply) => {
    const { providerName } = request.params;
    const staticConfig = getStaticConfig();
    const runtimeOverride = request.body || {};

    const checker = new ProviderHealthChecker(staticConfig, runtimeOverride);
    const result = await checker.check(providerName);

    return {
      provider: result.provider,
      status: result.status,
      latency_ms: result.latency_ms,
      message: result.message,
      error: result.error,
      timestamp: result.timestamp,
      config_used: result.config_used,
    };
  });

  fastify.log.info(`Provider connection routes registered at ${prefix}`);
}

export default fastifyPlugin(providerConnectionRoutesPlugin, {
  name: 'provider-connection-routes',
  fastify: '5.x',
});
